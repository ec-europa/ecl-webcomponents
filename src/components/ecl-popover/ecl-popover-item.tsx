import { Component, h, Prop } from '@stencil/core';

@Component({
  tag: 'ecl-popover-item',
  shadow: false,
  scoped: false,
})
export class EclPopoverItem {
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() path: string;
  @Prop() icon: string;
  @Prop() iconSprite: string = '';
  @Prop() iconSize: string = 'fluid';

  getClass(): string {
    const styleClasses = [
      `ecl-popover__item`,
      this.styleClass,
      `sc-ecl-popover-${this.theme}`,
    ];

    return styleClasses.join(' ');
  }

  getLinkClass(): string {
    const linkClasses = [
      'ecl-popover__link',
      `sc-ecl-popover-${this.theme}`,
    ];

    if (this.icon) {
      linkClasses.push('ecl-link--icon');
    }

    return linkClasses.join(' ');
  }

  render() {
    return (
      <li 
        class={this.getClass()}
      >
        <ecl-link
          style-class={`sc-ecl-popover-${this.theme} ecl-popover__link`}
          variant="standalone"
          path={this.path}
        >
        { this.icon ?
          <ecl-icon
            style-class={`sc-ecl-popover-${this.theme} ecl-link__icon`}
            icon={this.icon}
            slot="icon-before"
            sprite={this.iconSprite || ''}
            size={this.iconSize}
          ></ecl-icon> : '' 
        }
          <slot></slot>
        </ecl-link>
      </li>
    );
  }
}