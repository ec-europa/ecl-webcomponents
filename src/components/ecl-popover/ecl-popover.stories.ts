export default {
  title: 'Components/popover',
};

const Template = (args) => 
  `<ecl-popover
    ecl-script
    theme="${args.theme}"
    toggle-label="Popover"
    list
    icon="share"
    item-id="popover-example"
  >
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 1
    </ecl-popover-item>
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 2
    </ecl-popover-item>
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 3
    </ecl-popover-item>
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 4
    </ecl-popover-item>
      <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 5
    </ecl-popover-item>
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 6
    </ecl-popover-item>
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 7
    </ecl-popover-item>
    <ecl-popover-item
      path="/example.html"
      icon="global"
    >
      Item 8
    </ecl-popover-item>
  </ecl-popover>`;

export const Popover = Template.bind({});
Popover.storyName = 'default';