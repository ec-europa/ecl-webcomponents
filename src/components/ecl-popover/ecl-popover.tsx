import { Component, h, Prop, Element } from '@stencil/core';
import getAssetPath from "../../utils/assetPath";
declare const ECL: any;

@Component({
  tag: 'ecl-popover',
  styleUrls: {
    ec: './build/styles/ecl-popover-ec.css',
    eu: './build/styles/ecl-popover-eu.css',
  },
  shadow: false,
  scoped: true,
  assetsDirs: ['build'],
})
export class EclPopover {
  @Element() el: HTMLElement;
  @Prop() theme: string = 'ec';
  @Prop() styleClass: string;
  @Prop() eclScript: boolean = false;
  @Prop() itemId: string;
  @Prop() toggleLabel: string;
  @Prop() list: boolean = false;
  @Prop() icon: string;
  @Prop() iconSize: string = 'fluid';
  @Prop() iconSprite: string = '';

  getClass(): string {
    const styleClasses = [
      `ecl-popover`,
      this.styleClass
    ];

    return styleClasses.join(' ');
  }

  getAttrs() {
    return {
      'aria-controls': this.itemId,
      'data-ecl-popover-toggle': 'data-ecl-popover-toggle',
      'aria-expanded': 'false',
    };
  }

  componentDidLoad() {
    if (this.el.querySelector('.ecl-popover__list')) {
      // Clean the html so that the script finds what it expects.
      const items = this.el.querySelectorAll('.ecl-popover__item');
      this.el.querySelector('.ecl-popover__list').innerHTML = '';
      this.el.querySelector('.ecl-popover__list').append(...items);
    }
    if (this.eclScript) {
      // Load the ECL vanilla js if not already present.
      const src = getAssetPath('./build/scripts/ecl-popover-vanilla.js');
      if (document.querySelector(`script[src="${src}"]`)) {
        document.querySelector(`script[src="${src}"]`).remove();
      }
      const script = document.createElement('script');
      script.src = src;
      script.onload = () => {
        const popover = new ECL.Popover(this.el.firstElementChild);
        popover.init();
      };

      document.body.appendChild(script);
    }
  }

  render() {
    return (
      <div 
        class={this.getClass()}
      >
        <a
          class={`ecl-link sc-ecl-link-${this.theme} ecl-link--standalone ecl-popover__toggle`}
          {...this.getAttrs()}
        >
        { this.icon ?
          <ecl-icon
            theme={this.theme}
            icon={this.icon}
            size={this.iconSize}
            style-class="ecl-link__icon--before"
            sprite={this.iconSprite}
          ></ecl-icon> : ''
        }
          {this.toggleLabel}
        </a>
        <div
          class="ecl-popover__container"
          id={this.itemId}
          hidden
        >
          <div class="ecl-popover__content">
          { this.list ?
            <ul class="ecl-popover__list">
              <slot></slot>
            </ul> : ''
          }
          { !this.list ?
            <slot></slot> : ''
          }
          </div>
        </div>
      </div>
    );
  }
}