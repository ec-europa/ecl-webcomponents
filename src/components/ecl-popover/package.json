{
  "name": "@ecl/web-components-popover",
  "author": "European Commission",
  "license": "EUPL-1.2",
  "version": "0.11.0",
  "description": "ECL Popover",
  "main": "ecl-popover.tsx",
  "private": true,
  "scripts": {
    "build:scripts": "cross-env ecl-builder scripts",
    "build:styles": "cross-env ecl-builder styles",
    "build": "run-s clean build:*",
    "clean": "rimraf build",
    "prepublish": "npm run build"
  },
  "devDependencies": {
    "@ecl/builder": "3.10.0",
    "@ecl/vanilla-component-link": "3.10.0",
    "@ecl/vanilla-component-popover": "3.10.0",
    "cross-env": "7.0.3"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planctus/ecl-webcomponents.git"
  },
  "bugs": {
    "url": "https://github.com/planctus/ecl-webcomponents/issues"
  },
  "homepage": "https://github.com/planctus/ecl-webcomponents",
  "keywords": [
    "ecl",
    "europa-component-library",
    "design-system",
    "web-components"
  ]
}
