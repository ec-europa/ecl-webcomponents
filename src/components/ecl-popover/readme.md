# ecl-popover-item



<!-- Auto Generated Below -->


## Properties

| Property     | Attribute     | Description | Type     | Default     |
| ------------ | ------------- | ----------- | -------- | ----------- |
| `icon`       | `icon`        |             | `string` | `undefined` |
| `iconSize`   | `icon-size`   |             | `string` | `'fluid'`   |
| `iconSprite` | `icon-sprite` |             | `string` | `''`        |
| `path`       | `path`        |             | `string` | `undefined` |
| `styleClass` | `style-class` |             | `string` | `undefined` |
| `theme`      | `theme`       |             | `string` | `'ec'`      |


## Dependencies

### Depends on

- [ecl-link](../ecl-link)
- [ecl-icon](../ecl-icon)

### Graph
```mermaid
graph TD;
  ecl-popover-item --> ecl-link
  ecl-popover-item --> ecl-icon
  ecl-link --> ecl-icon
  style ecl-popover-item fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
