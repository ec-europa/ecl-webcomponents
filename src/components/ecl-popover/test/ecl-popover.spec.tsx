import { newSpecPage } from '@stencil/core/testing';
import { EclPopover } from '../ecl-popover';
import { EclPopoverItem } from '../ecl-popover-item';

describe('ecl-popover', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [EclPopover, EclPopoverItem],
      html:   `<ecl-popover
        ecl-script
        theme="ec"
        toggle-label="Popover"
        list
        icon="share"
        item-id="popover-example"
      >
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 1
        </ecl-popover-item>
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 2
        </ecl-popover-item>
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 3
        </ecl-popover-item>
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 4
        </ecl-popover-item>
          <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 5
        </ecl-popover-item>
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 6
        </ecl-popover-item>
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 7
        </ecl-popover-item>
        <ecl-popover-item
          path="/example"
          icon="global"
        >
          Item 8
        </ecl-popover-item>
      </ecl-popover>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
