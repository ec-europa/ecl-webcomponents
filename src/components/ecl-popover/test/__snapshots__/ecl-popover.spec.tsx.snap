// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecl-popover renders 1`] = `
<ecl-popover ecl-script="" icon="share" item-id="popover-example" list="" theme="ec" toggle-label="Popover">
  <!---->
  <div class="ecl-popover">
    <a aria-controls="popover-example" aria-expanded="false" class="ecl-link ecl-link--standalone ecl-popover__toggle" data-ecl-popover-toggle="data-ecl-popover-toggle">
      <ecl-icon icon="share" size="fluid" style-class="ecl-link__icon--before"></ecl-icon>
      Popover
    </a>
    <div class="ecl-popover__container" hidden="" id="popover-example">
      <div class="ecl-popover__content">
        <ul class="ecl-popover__list">
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 1
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 2
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 3
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 4
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 5
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 6
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 7
            </ecl-link>
          </li>
          <li class="ecl-popover__item sc-ecl-popover-ec">
            <ecl-link path="/example.html" style-class="sc-ecl-popover-ec ecl-popover__link" variant="standalone">
              <ecl-icon icon="global" size="fluid" slot="icon-before" style-class="sc-ecl-popover-ec ecl-link__icon"></ecl-icon>
              Item 8
            </ecl-link>
          </li>
        </ul>
      </div>
    </div>
  </div>
</ecl-popover>
`;
